File: Data Structures & Algorithms/binary-tree-inorder-traversal/submission-2.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun inorderTraversal(root: TreeNode?): List<Int> {
        val result = mutableListOf<Int>()
        val stack = ArrayDeque<TreeNode>()
        var current = root

        while (current != null || stack.isNotEmpty()) {

            // Go all the way left
            while (current != null) {
                stack.addLast(current)
                current = current.left
            }

            // Process node
            current = stack.removeLast()
            result.add(current.`val`)

            // Move to right subtree
            current = current.right
        }

        return result
    }
}

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-2.cpp
// BFS
class Solution {
   public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        if (root == nullptr) return result;

        queue<TreeNode*> queue;
        queue.push(root);

        while (!queue.empty()) {
            vector<int> level;
            int size = queue.size();

            for (int i = size; i > 0; i--) {
                TreeNode* node = queue.front();
                queue.pop();

                if (node != nullptr) {
                    level.push_back(node->val);
                    if (node->left) queue.push(node->left);
                    if (node->right) queue.push(node->right);
                }
            }
            if (!level.empty()) {
                result.push_back(level);
            }
        }

        return result;
    }
};
